x/participationrewards/keeper/hooks.go

```diff
@@ -146,6 +150,120 @@ func (k *Keeper) AfterZoneCreated(ctx sdk.Context, zone *icstypes.Zone) error {
+	if !found {
+		return fmt.Errorf("channel not found: %s", zone.TransferChannel)
+	}
+
+	// Create LiquidAllowedDenomProtocolData for the host zone
+	hostZoneDenom := types.LiquidAllowedDenomProtocolData{
+		ChainID:               zone.ChainId,
+		RegisteredZoneChainID: zone.ChainId,
+		IbcDenom:              utils.DeriveIbcDenom(transfertypes.PortID, channel.Counterparty.ChannelId, transfertypes.PortID, zone.TransferChannel, zone.LocalDenom),
+		QAssetDenom:           zone.LocalDenom,
+	}
+	if err := hostZoneDenom.ValidateBasic(); err != nil {
+		return err
+	}
+	hostZoneDenomBytes, err := json.Marshal(hostZoneDenom)
+	if err != nil {
+		return err
+	}
+	k.SetProtocolData(ctx, hostZoneDenom.GenerateKey(), &types.ProtocolData{
+		Type: types.ProtocolDataType_name[int32(types.ProtocolDataTypeLiquidToken)],
+		Data: hostZoneDenomBytes,
+	})
+
+	// Fetch OsmosisParamsProtocolData and create LiquidAllowedDenomProtocolData for Osmosis
+	k.IteratePrefixedProtocolDatas(ctx, types.GetPrefixProtocolDataKey(types.ProtocolDataTypeOsmosisParams), func(index int64, _ []byte, data types.ProtocolData) (stop bool) {
+		osmosisParams, err := types.UnmarshalProtocolData(types.ProtocolDataTypeOsmosisParams, data.Data)
+		if err != nil {
+			return false
+		}
+		osmosisParamsData, ok := osmosisParams.(*types.OsmosisParamsProtocolData)
+		if !ok {
+			return false
+		}
+		osmosisConnectionData, found := k.GetProtocolData(ctx, types.ProtocolDataTypeConnection, osmosisParamsData.ChainID)
+		if !found {
+			return false
+		}
+		connectionData, err := types.UnmarshalProtocolData(types.ProtocolDataTypeConnection, osmosisConnectionData.Data)
+		if err != nil {
+			return false
+		}
+		osmosisChannel := connectionData.(*types.ConnectionProtocolData).TransferChannel
+
+		// channel for the osmosis chain
+		channel, found := k.IBCKeeper.ChannelKeeper.GetChannel(ctx, "transfer", osmosisChannel)
+		if !found {
+			return false
+		}
+		osmosisDenom := types.LiquidAllowedDenomProtocolData{
+			ChainID:               osmosisParamsData.ChainID,
+			RegisteredZoneChainID: zone.ChainId,
+			IbcDenom:              utils.DeriveIbcDenom(transfertypes.PortID, channel.Counterparty.ChannelId, transfertypes.PortID, osmosisChannel, zone.LocalDenom),
+			QAssetDenom:           zone.LocalDenom,
+		}
+		if err := osmosisDenom.ValidateBasic(); err != nil {
+			return false
+		}
+		osmosisDenomBytes, err := json.Marshal(osmosisDenom)
+		if err != nil {
+			return false
+		}
+		k.SetProtocolData(ctx, osmosisDenom.GenerateKey(), &types.ProtocolData{
+			Type: types.ProtocolDataType_name[int32(types.ProtocolDataTypeLiquidToken)],
+			Data: osmosisDenomBytes,
+		})
+		return true
+	})
+
+	// Fetch UmeeParamsProtocolData and create LiquidAllowedDenomProtocolData for Umee
+	k.IteratePrefixedProtocolDatas(ctx, types.GetPrefixProtocolDataKey(types.ProtocolDataTypeUmeeParams), func(index int64, _ []byte, data types.ProtocolData) (stop bool) {
```

Comment:
Same here; just GetProtocolData - key is constant.

Reply:
Also, can't help thinking reading this, that `GetAndUnmarshalProtocolData() would be cleaner here?

And maybe a corresponding `MarshalAndSetProtocolData()`.

Perhaps using generics if needed.